numpy/core/src/multiarray/descriptor.c

```diff
@@ -1072,7 +1072,13 @@ _convert_from_dict(PyObject *obj, int align)
                 goto fail;
             }
             offset = PyInt_AsLong(off);
-            PyTuple_SET_ITEM(tup, 1, off);
+            if (PyErr_Occurred()) {
```

Comment:
`if ((offset == -1) && PyErr_Occured()) {` avoids the check most of the time I think.